ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -95,9 +131,27 @@ private[kvutils] object InputsAndEffects {
               l.result.fold(inputs)(inputs ++ contractInputs(_)) +
                 contractKeyToStateKey(GlobalKey(l.templateId, forceNoContractIds(l.key.key)))
           }
+      }.toList ++ partyInputsInValues()
+    }
+
+  def subValues(v: Value[ContractId]): ImmArray[Value[ContractId]] =
```

Comment:
Seems wasteful to construct a temporary data structure here. I would consider using local mutability for computing the party inputs (e.g. have a function to traverse into transactions and their values and maintain a mutable set in lexical scope). The overly functional style is unfortunately too costly since we might have massive transactions to deal with and have to avoid unnecessary allocation as much as possible.

Reply:
OK, as agreed verbally I'll go mutable there not only regarding party inputs but inputs in general (makes sense to have this refactoring here even though it's slightly out-of-scope).